src/gui/torrentcontenttreeview.cpp

```diff
@@ -75,6 +87,239 @@ void TorrentContentTreeView::keyPressEvent(QKeyEvent *event)
+        }
+
+        // check if that name is already used
+        for (int i = 0; i < torrent->filesCount(); ++i) {
+            if (i == fileIndex) continue;
+            if (Utils::Fs::sameFileNames(torrent->filePath(i), newFilePath)) {
+                RaisedMessageBox::warning(this, tr("Rename error"),
+                                          tr("This name is already in use in this folder. Please use a different name."),
+                                          QMessageBox::Ok);
+                return;
+            }
+        }
+
+        qDebug("Renaming %s to %s", qUtf8Printable(oldFilePath), qUtf8Printable(newFilePath));
+        torrent->renameFile(fileIndex, newFilePath);
+
+        model->setData(modelIndex, newName);
+    }
+    else {
+        // renaming a folder
+
+        const QString oldName = modelIndex.data().toString();
+        if (newName == oldName)
+            return;  // Name did not change
+
+        QString parentPath;
+        for (QModelIndex idx = model->parent(modelIndex); idx.isValid(); idx = model->parent(idx))
+            parentPath.prepend(idx.data().toString() + '/');
+
+        const QString oldPath {parentPath + oldName + '/'};
+        const QString newPath {parentPath + newName + '/'};
+
+        // Check for overwriting
+        for (int i = 0; i < torrent->filesCount(); ++i) {
+            const QString currentPath = torrent->filePath(i);
+
+            if (currentPath.startsWith(oldPath))
+                continue;
+
+#if defined(Q_OS_WIN)
+            const Qt::CaseSensitivity caseSensitivity = Qt::CaseInsensitive;
+#else
+            const Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive;
+#endif
+            if (currentPath.startsWith(newPath, caseSensitivity)) {
+                RaisedMessageBox::warning(this, tr("The folder could not be renamed"),
+                                          tr("This name is already in use. Please use a different name."),
+                                          QMessageBox::Ok);
+                return;
+            }
+        }
+
+        // Replace path in all files
+        bool needForceRecheck = false;
+
+        for (int i = 0; i < torrent->filesCount(); ++i) {
+            const QString currentPath = torrent->filePath(i);
+
+            if (currentPath.startsWith(oldPath)) {
+                const QString path {newPath + currentPath.mid(oldPath.length())};
+
+                if (!needForceRecheck && QFile::exists(path))
+                    needForceRecheck = true;
+
+                torrent->renameFile(i, path);
+            }
+        }
+
+        // Force recheck
+        if (needForceRecheck)
+            torrent->forceRecheck();
+
+        // Remove old folder
+        const QString oldFullPath = torrent->savePath(true) + oldPath;
+        int timeout = 10;
+        while (!QDir().rmpath(oldFullPath) && (timeout > 0)) {
+            // FIXME: We should not sleep here (freezes the UI for 1 second)
+            QThread::msleep(100);
```

Comment:
@Chocobo1:  While you are working on this code, perhaps this old `FIXME` issue can be resolved?

Reply:
I might resolve it in another PR.